### PR TITLE
Add Async Storage shutdown

### PR DESCRIPTION
**Pull request by 0xTim**
Currently running

```swift
Task {
    try? await Task.sleep(for: .seconds(5))
    app.running?.stop()
}
```

When you try and install NIO as the global executor will crash because the storage API didn't have any async entry points so stopping would trigger a synchronous shutdown, with a `wait()`. This fixes that

### CODE REVIEW

File: Sources/Vapor/Commands/ServeCommand.swift [hunk 1]
```diff
@@ -115,6 +116,18 @@ public final class ServeCommand: AsyncCommand, Sendable {
         self.box.withLockedValue { $0 = box }
     }
     
+    func asyncShutdown() async {
+        var box = self.box.withLockedValue { $0 }
+        box.didShutdown = true
+        box.running?.stop()
+        if let server = box.server {
+            await server.shutdown()
+        }
```
Comment (gwynne) [inline Sources/Vapor/Commands/ServeCommand.swift:125]:
```suggestion
        await box.server?.shutdown()
```

File: Sources/Vapor/Utilities/Storage.swift [hunk 2]
```diff
@@ -11,13 +11,21 @@ public struct Storage: Sendable {
...
                 logger.warning("Could not shutdown \(T.self): \(error)")
             }
         }
+        func asyncShutdown(logger: Logger) async {
+            do {
+                try await self.onAsyncShutdown?(self.value)
+            } catch {
+                logger.warning("Could not shutdown \(T.self): \(error)")
+            }
+        }
```
Comment (gwynne) [inline Sources/Vapor/Utilities/Storage.swift:28]: I don't have a better suggestion, but this seems incredibly error-prone/easy to misuse 😕

File: Sources/Vapor/Utilities/Storage.swift [hunk 3]
```diff
@@ -94,20 +102,49 @@ public struct Storage: Sendable {
...
+    ) async
+        where Key: StorageKey
+    {
+        let key = ObjectIdentifier(Key.self)
+        if let value = value {
+            self.storage[key] = Value(value: value, onShutdown: nil, onAsyncShutdown: onShutdown)
+        } else if let existing = self.storage[key] {
+            self.storage[key] = nil
+            await existing.asyncShutdown(logger: self.logger)
+        }
```
Comment (gwynne) [inline Sources/Vapor/Utilities/Storage.swift:122]: What if the existing had a sync shutdown handler but no async one?

**Review by gwynne — APPROVED** [review]
A nit and a question, but nothing blocking. I find this _incredibly_ ugly and kinda concerningly fragile, but I have no better suggestions 😕

File: Sources/Vapor/HTTP/Client/Application+HTTP+Client.swift [hunk 4]
```diff
@@ -34,6 +35,28 @@ extension Application.HTTP {
                 return new
             }
         }
+        
+        public var asyncShared: HTTPClient {
+            get async {
+                let lock = self.application.locks.lock(for: Key.self)
+                lock.lock()
```
Comment (0xTim) [inline Sources/Vapor/HTTP/Client/Application+HTTP+Client.swift:42]: This makes me uneasy but I don't see a way around it - any thoughts?

File: Sources/Vapor/Utilities/Storage.swift (hunk 3 above)
Comment (0xTim) [reply Sources/Vapor/Utilities/Storage.swift:124]: This is handled in `asyncShutdown()` - it will run both

**Comment by 0xTim** [comment]
Ok @gwynne I've tested this with client calls as well and think I've fixed all the different areas without breaking the API too much

File: Sources/Vapor/Utilities/Storage.swift (hunk 3 above)
Comment (0xTim) [reply Sources/Vapor/Utilities/Storage.swift:127]: Ok now it will run the sync one if an async one doesn't exist

**Review by gwynne — APPROVED** [review]

**Pull request merged**